training/CHAPEL_TUTORIAL.md

```diff
@@ -993,358 +1073,635 @@ Expected Output:
+  src[0] is remote
+  dst[3] is local
+
+iteration num: 5
+  on Locale: 1
+  on localSubDom: {3..5}
+  src[2] is remote
+  dst[5] is local
+
+iteration num: 4
+  on Locale: 1
+  on localSubDom: {3..5}
+  src[1] is remote
+  dst[4] is local
+
+src: 0 1 2 3 4 5
+dst: 3 4 5 0 1 2
+```
 
-var upperEntry = new SymEntry(SIZE, uint);
-var lowerEntry = new SymEntry(SIZE, uint);
+<a id="perf"></a>
+## Performance and Diagnostics
+Now that you're a pro at writing chapel code, let's talk about how to make the code you write more efficient!
 
-resetCommDiagnostics();
-startCommDiagnostics();
-var t1 = Time.timeSinceEpoch().totalSeconds();
+<a id="var_dec"></a>
+### Variable Declarations
+Throughout this tutorial we've mainly used `var` to declare variables, but this is not the only option.
+The driving factor of using other declarations is to improve performance or reduce memory usage.
 
-// the leading iterator is a range, so all the computation happens on locale 0
-forall (i, (up, low)) in zip(0..#SIZE, hashes) {
-  upperEntry.a[i] = up;
-  lowerEntry.a[i] = low;
-}
+[`const`](https://chapel-lang.org/docs/users-guide/base/constParam.html):
+* can be used when we know a variable should never be changed
+* a program execution time constant. This means the compiler doesn't need to know its value, and it can be 
+set at runtime (i.e. passed in by a user)
+* knowing this value won't change can help the compiler to make optimizations
 
-var t2 = Time.timeSinceEpoch().totalSeconds() - t1;
-stopCommDiagnostics();
-writeln("leading iterator not distributed time = ", t2);
-writeln();
-printCommDiagnosticsTable();
+[`param`](https://chapel-lang.org/docs/users-guide/base/constParam.html#declaring-params):
+* similar to `const`, except it must be known at compile time.
 
-resetCommDiagnostics();
-startCommDiagnostics();
-t1 = Time.timeSinceEpoch().totalSeconds();
+[`ref`](https://chapel-lang.org/docs/language/spec/variables.html?highlight=ref#ref-variables):
+* creates a reference to an existing object
+* avoids creating a copy (less memory use)
+* any updates to this will update the object it is referencing 
+* useful to avoid repeated accesses of an attribute (especially within a loop)
 
-// leading iterator is distributed
-// but every iteration access the `.a` component
-forall (i, (up, low)) in zip(hashes.domain, hashes) {
-  upperEntry.a[i] = up;
-  lowerEntry.a[i] = low;
-}
 
-t2 = Time.timeSinceEpoch().totalSeconds() - t1;
-stopCommDiagnostics();
-writeln("leading iterator is distributed, but .a accesses time = ", t2);
-writeln();
-printCommDiagnosticsTable();
+[`const ref`](https://chapel-lang.org/docs/language/spec/variables.html?highlight=ref#ref-variables)
+* same as a `ref` but cannot be changed
+
+<a id="diagnostics"></a>
+### Diagnostics
+
+There are several chapel modules available to aid in optimizing your code. The most common ones
+used by the arkouda team are [comm](https://chapel-lang.org/docs/modules/standard/CommDiagnostics.html)
```

Comment:
Maybe 'CommDiagnostics'?  (and 'Time' rather than 'time'?)

Or change "are comm and time" to "measure communication and time"?